### PR TITLE
Fix gss_add_cred() crash when output_cred_handle is NULL (fix #413)

### LINKED ISSUE #413: gss_add_cred() crashes on null output_cred_handle
RFC 2744 says about gss_add_cred():
```
   This routine can be used to either compose a new credential
   containing all credential-elements of the original in addition to the
   newly-acquire credential-element, or to add the new credential-
   element to an existing credential. If NULL is specified for the
   output_cred_handle parameter argument, the new credential-element
   will be added to the credential identified by input_cred_handle; if a
   valid pointer is specified for the output_cred_handle parameter, a
   new credential handle will be created.
```
Heimdal's mechglue gss_add_cred() does not expect output_cred_handle to be NULL; it dereferences it unconditionally at the beginning and end of the function, and only contains logic to create a new credential (with or without a copy of mechanism creds from an old credential), not to mutate an existing one.

(I discovered this through code inspection while looking into bugs in MIT krb5's version of the same function.  I have not tried to reproduce it, and I don't believe gss_add_cred() is used much in the wild.)

### PR DESCRIPTION
**Pull request by nicowilliams**
First pass.  I'm thinking of removing the `target_mc` stuff -- if you try to add an element for a mech you already have an element for then we should return `GSS_S_COMPLETE` and be done.

### CODE REVIEW

**Comment by nicowilliams** [comment]
Heimdal's interpretation of GSS is a bit funky:

- We allow `gss_name_t` instances to have multiple mech names (MNs).  I suspect this is about SPNEGO, specifically about inquiring a SPNEGO credential's name, but it really shouldn't be necessary.

- We seem to allow `gss_cred_id_t` instances to have multiple `{mech, name}` elements.  This can be good for acceptors, but it's pretty pointless for initiators.  Acceptors should just always use the default credential and then inquire the context for the name used by the initiator and possibly reject (at the application layer) if that name was not acceptable.

**Comment by lhoward** [comment]
did the mcache fix accidentally slip in here?

File: lib/gssapi/mech/gss_add_cred.c [hunk 1]
```diff
@@ -93,94 +93,132 @@ gss_add_cred(OM_uint32 *minor_status,
...
+            }
+            return GSS_S_COMPLETE;
+        }
+
+
+        if (output_cred_handle) {
+            new_cred = malloc(sizeof(struct _gss_cred));
+            if (!new_cred) {
+                *minor_status = ENOMEM;
+                return (GSS_S_FAILURE);
```
Comment (lhoward) [inline lib/gssapi/mech/gss_add_cred.c]: return GSS_S_FAILURE

Comment (lhoward) [reply lib/gssapi/mech/gss_add_cred.c]: picky I know

Comment (nicowilliams) [reply lib/gssapi/mech/gss_add_cred.c]: Oh, I agree.

**Comment by nicowilliams** [comment]
Which mcache fix?  The one for the deadlock?  I pushed it.

File: lib/gssapi/mech/gss_add_cred.c (hunk 1 above)
Comment (nicowilliams) [reply lib/gssapi/mech/gss_add_cred.c]: Pushed a separate commit to restyle this file.

**Comment by nicowilliams** [comment]
The Travis OS X build evinced a `check-ldap` failure.  I don't know why that would be.  I've restarted that build in case it's spurious.

**Comment by nicowilliams** [comment]
The `check-ldap` failure is due to a race condition in the test that I'll fix separately.

**Comment by nicowilliams** [comment]
Needs more work.

File: lib/gssapi/mech/gss_add_cred.c [hunk 2]
```diff
@@ -83,104 +84,120 @@ gss_add_cred(OM_uint32 *minor_status,
...
+        return GSS_S_BAD_MECH;
+    if (input_cred_handle == GSS_C_NO_CREDENTIAL &&
+        output_cred_handle == NULL) {
+        return GSS_S_CALL_INACCESSIBLE_WRITE;
+    }
+
+    /* Setup mut_cred to be the credential we mutate */
+    if (input_cred_handle != GSS_C_NO_CREDENTIAL &&
+        output_cred_handle != NULL) {
+        gss_cred_id_t new_cred = GSS_C_NO_CREDENTIAL;
```
Comment (nicowilliams) [inline lib/gssapi/mech/gss_add_cred.c]: Initialization not needed here.

File: lib/gssapi/mech/gss_add_cred.c [hunk 3]
```diff
@@ -83,104 +84,120 @@ gss_add_cred(OM_uint32 *minor_status,
...
+                                        (gss_const_cred_id_t)mut_cred, NULL,
+                                        NULL, NULL, actual_mechs);
+        if (major_status != GSS_S_COMPLETE)
+            _gss_mg_error(m, major_status, *minor_status);
+    }
+    if (major_status == GSS_S_COMPLETE)
+        *output_cred_handle = (gss_cred_id_t)mut_cred;
+    else
+        gss_release_cred(&junk, &release_cred);
+    free(mc);
```
Comment (nicowilliams) [inline lib/gssapi/mech/gss_add_cred.c]: s/mc/new_mc/

File: lib/gssapi/netlogon/duplicate_cred.c [hunk 4]
```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright (c) 2010 Kungliga Tekniska Högskolan
```
Comment (nicowilliams) [inline lib/gssapi/netlogon/duplicate_cred.c]: Wrong copyright.

File: lib/gssapi/ntlm/duplicate_cred.c [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    new_cred->key.data = malloc(cred->key.length);
+    if (new_cred->username == NULL || new_cred->domain == NULL ||
+        new_cred->key.data == NULL) {
+        *output_cred_handle = (gss_cred_id_t) new_cred;
+        _gss_ntlm_release_cred(&junk, output_cred_handle);
+        *minor_status = ENOMEM;
+        return GSS_S_FAILURE;
+    }
+
+    new_cred->key.length = cred->key.length;
```
Comment (nicowilliams) [inline lib/gssapi/ntlm/duplicate_cred.c]: Also copy the key contents!

File: lib/gssapi/ntlm/duplicate_cred.c [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    OM_uint32 junk;
+
+    *output_cred_handle = GSS_C_NO_CREDENTIAL;
+    if ((new_cred = calloc(1, sizeof(*new_cred))) == NULL) {
+        *minor_status = ENOMEM;
+        return GSS_S_FAILURE;
+    }
+    
+    new_cred->usage = cred->usage;
+    new_cred->username = strdup(cred->username);
```
Comment (nicowilliams) [inline lib/gssapi/ntlm/duplicate_cred.c]: What if the input cred is `GSS_C_NO_CREDENTIAL`?

**Pull request merged**